services/local-notifications/core.go

```diff
@@ -0,0 +1,338 @@
+				pushMessage(n)
+			}
+		}
+	}
+}
+
+// SubscribeWallet - Subscribes to wallet signals
+func (s *Service) SubscribeWallet(publisher *event.Feed) error {
+	s.walletTransmitter.publisher = publisher
+
+	preference, err := s.db.GetWalletPreference()
+
+	if err != nil {
+		log.Error("Failed to get wallet preference", "error", err)
+		return nil
+	}
+
+	if preference.Enabled {
+		s.StartWalletWatcher()
+	}
+
+	return nil
+}
+
+// StartWalletWatcher - Forward wallet events to notifications
+func (s *Service) StartWalletWatcher() {
+	if s.walletTransmitter.quit != nil {
+		// already running, nothing to do
+		return
+	}
+
+	if s.walletTransmitter.publisher == nil {
+		log.Error("wallet publisher was not initialized")
+		return
+	}
+
+	s.walletTransmitter.quit = make(chan struct{})
+	events := make(chan wallet.Event, 10)
+	sub := s.walletTransmitter.publisher.Subscribe(events)
+
+	s.walletTransmitter.wg.Add(1)
+
+	go func() {
+		defer s.walletTransmitter.wg.Done()
+		for {
+			select {
+			case <-s.walletTransmitter.quit:
+				sub.Unsubscribe()
+				return
+			case err := <-sub.Err():
+				// technically event.Feed cannot send an error to subscription.Err channel.
+				// the only time we will get an event is when that channel is closed.
+				if err != nil {
+					log.Error("wallet signals transmitter failed with", "error", err)
+				}
+				return
+			case event := <-events:
+
+				if event.Type == wallet.EventNewBlock {
+					s.transmitter.publisher.Send(TransactionEvent{
+						Type:                      string(event.Type),
+						BlockNumber:               event.BlockNumber,
+						Accounts:                  []common.Address(event.Accounts),
+						NewTransactionsPerAccount: map[common.Address]int(event.NewTransactionsPerAccount),
+						ERC20:                     bool(event.ERC20),
+					})
+				}
+			}
+		}
+	}()
+}
+
+// StopWalletWatcher - stops watching for new wallet events
+func (s *Service) StopWalletWatcher() {
+	if s.walletTransmitter.quit != nil {
+		close(s.walletTransmitter.quit)
+		s.walletTransmitter.wg.Wait()
+		s.walletTransmitter.quit = nil
+	}
+}
+
+// IsWatchingWallet - check if local-notifications are subscribed to wallet updates
+func (s *Service) IsWatchingWallet() bool {
+	return s.walletTransmitter.quit != nil
+}
+
+// Start Worker which processes all incoming messages
+func (s *Service) Start(_ *p2p.Server) error {
+	s.started = true
+
+	s.transmitter.quit = make(chan struct{})
+	s.transmitter.publisher = &event.Feed{}
+
+	events := make(chan TransactionEvent, 10)
+	sub := s.transmitter.publisher.Subscribe(events)
+
+	s.transmitter.wg.Add(1)
+	go func() {
+		defer s.transmitter.wg.Done()
+		for {
+			select {
+			case <-s.transmitter.quit:
+				sub.Unsubscribe()
+				return
+			case err := <-sub.Err():
+				if err != nil {
+					log.Error("Local notifications transmitter failed with", "error", err)
+				}
+				return
+			case event := <-events:
+				s.transactionsHandler(event)
+			}
+		}
+	}()
```

Comment:
Same for this code. If the parent function's progress is blocked by the subroutine and no other work is performed, do we need to have a subroutine?